zk-token-sdk/src/curve25519/curve_syscall_traits.rs

```diff
@@ -0,0 +1,98 @@
+//! The traits representing the basic elliptic curve operations.
+//!
+//! These traits are instantiatable by all the commonly used elliptic curves and should help in
+//! organizing syscall support for other curves in the future. more complicated or curve-specific
+//! functions that are needed in cryptographic applications should be representable by combining
+//! the associated functions of these traits.
+//!
+//! NOTE: This module temporarily lives in zk_token_sdk/curve25519, but it is independent of
+//! zk-token-sdk or curve25519. It should be moved to a more general location in the future.
+//!
+
+pub trait PointValidation {
+    type Point;
+
+    /// Verifies if a byte representation of a curve point lies in the curve.
+    fn validate_point(&self) -> bool;
+}
+
+pub trait GroupOperations {
+    type Point;
+    type Scalar;
+
+    /// Adds two curve points: P_0 + P_1.
+    fn add(left_point: &Self::Point, right_point: &Self::Point) -> Option<Self::Point>;
+
+    /// Subtracts two curve points: P_0 - P_1.
+    ///
+    /// NOTE: Altneratively, one can consider replacing this with a `negate` function that maps a
+    /// curve point P -> -P. Then subtraction can be computed by combining `negate` and `add`
+    /// syscalls. However, `subtract` is a much more widely used function than `negate`.
+    fn subtract(left_point: &Self::Point, right_point: &Self::Point) -> Option<Self::Point>;
+
+    /// Multiplies a scalar S with a curve point P: S*P
+    fn multiply(scalar: &Self::Scalar, point: &Self::Point) -> Option<Self::Point>;
+}
```

Comment:
I _think_ these can be implemented fully in terms of the associated types, just once here

Reply:
Since we are instantiating these traits for `PodEdwardsPoint` and `PodRistrettoPoint` structs, which don't instantiate `Add`, `Sub`, or `Mul`, it seems like we cannot implement these just with associated types. We can implement `Add`, `Sub`, and `Mul` for the individual pod structs, but it seems like this doesn't buy us much in terms of code organization/simplification.